File: plugins/ekey_flutter_sdk/android/src/main/kotlin/com/nec/ekey_flutter_sdk/EkeyFlutterSdkPlugin.kt
package com.nec.ekey_flutter_sdk

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.example.ekeysdk.Ekey
import com.example.ekeysdk.EkeyLoginResult
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONObject

/**
 * Flutter bridge to the bundled native `EkeySDK.aar`.
 *
 * Method channel `ekey_flutter_sdk`:
 *   - `initiateLogin`     → starts the flow, resolves with `{status, redirectUri?, ...}`
 *   - `getPendingResult`  → returns + clears a result that arrived while no Dart
 *                           call was waiting (e.g. after the host process / Activity
 *                           was killed while the eKey app was in the foreground)
 *
 * Event channel `ekey_flutter_sdk/events` streams the same result maps, and replays
 * a persisted-but-unconsumed result to the first listener.
 *
 * ### Why the persistence
 * EkeySDK delivers its result through an in-memory callback on `Ekey`. When Android
 * kills the host Activity (and, with the default `FlutterActivity`, its engine +
 * Dart isolate) while the user is in the eKey app, that callback fires into a dead
 * isolate and the `initiateLogin()` Future is lost. So every result is also written
 * to `SharedPreferences` and can be recovered after the app comes back.
 */
class EkeyFlutterSdkPlugin : FlutterPlugin, ActivityAware, MethodCallHandler,
    EventChannel.StreamHandler {

    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var activity: Activity? = null

    /** Method result for an in-flight `initiateLogin()` call on THIS engine. */
    private var pendingResult: Result? = null
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        appContext = binding.applicationContext
        instance = this

        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(binding.binaryMessenger, "$CHANNEL/events")
        eventChannel.setStreamHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        if (instance === this) instance = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initiateLogin" -> initiateLogin(result)
            "getPendingResult" -> result.success(takePersistedResult(appContext))
            else -> result.notImplemented()
        }
    }

    private fun initiateLogin(result: Result) {
        val current = activity
        if (current == null) {
            result.error(
                "NO_ACTIVITY",
                "No foreground Activity to present the eKey login screen from",
                null,
            )
            return
        }

        // Replace any earlier waiter (shouldn't happen; be safe).
        pendingResult?.let { runCatching { it.error("CANCELLED", "Superseded by a new login", null) } }
        pendingResult = result

        Ekey.initiateLogin(current) { loginResult -> deliver(loginResult.toMap()) }
    }

    // MARK: EventChannel.StreamHandler

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
        // Replay a result that landed while nothing was listening.
        takePersistedResult(appContext)?.let { events?.success(it) }
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    // MARK: ActivityAware

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    private companion object {
        const val CHANNEL = "ekey_flutter_sdk"
        const val PREFS = "ekey_flutter_sdk"
        const val KEY_RESULT = "pending_result"

        private val mainHandler = Handler(Looper.getMainLooper())

        /** Set while any engine of this process is attached. */
        @Volatile
        private var instance: EkeyFlutterSdkPlugin? = null

        @Volatile
        private var appContext: Context? = null

        /**
         * Called by the `Ekey` callback — which may outlive the engine/isolate that
         * started the flow. Persists first (survives process death), then delivers to
         * whatever Dart side is alive now.
         */
        private fun deliver(map: HashMap<String, Any?>) {
            appContext?.let { persistResult(it, map) }
            mainHandler.post {
                val plugin = instance
                val waiter = plugin?.pendingResult
                val sink = plugin?.eventSink
                if (waiter == null && sink == null) return@post // stays persisted for recovery

                takePersistedResult(appContext) // consume
                if (waiter != null) {
                    plugin.pendingResult = null
                    runCatching { waiter.success(map) }
                }
                runCatching { sink?.success(map) }
            }
        }

        private fun EkeyLoginResult.toMap(): HashMap<String, Any?> {
            val map = HashMap<String, Any?>()
            when (this) {
                is EkeyLoginResult.Completed -> {
                    map["status"] = "completed"
                    map["redirectUri"] = redirectUri.toString()
                    map["codeVerifier"] = codeVerifier
                    // EKYC payload — EkeySDK does the token exchange internally.
                    map["claims"] = sanitize(identity.claims)
                    identity.kycData?.let { map["kycData"] = sanitize(it) }
                }
                is EkeyLoginResult.Failed -> {
                    map["status"] = "failed"
                    map["error"] = error.toString()
                }
                else -> map["status"] = "cancelled" // EkeyLoginResult.Cancelled
            }
            return map
        }

        /** Coerce the claims/kyc tree to types the Flutter StandardMessageCodec accepts. */
        private fun sanitize(value: Any?): Any? = when (value) {
            null -> null
            is String, is Boolean, is Int, is Long, is Double -> value
            is Float -> value.toDouble()
            is Number -> value.toDouble()
            is Map<*, *> -> value.entries.associate { (k, v) -> k.toString() to sanitize(v) }
            is Iterable<*> -> value.map { sanitize(it) }
            else -> value.toString()
        }

        private fun persistResult(context: Context, map: Map<String, Any?>) {
            val json = JSONObject()
            for ((k, v) in map) json.put(k, v)
            context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
                .edit().putString(KEY_RESULT, json.toString()).apply()
        }

        /** Returns the persisted result map (and deletes it), or null. */
        private fun takePersistedResult(context: Context?): Map<String, Any?>? {
            context ?: return null
            val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            val raw = prefs.getString(KEY_RESULT, null) ?: return null
            prefs.edit().remove(KEY_RESULT).apply()
            return runCatching {
                val json = JSONObject(raw)
                buildMap<String, Any?> {
                    for (key in json.keys()) put(key, json.opt(key)?.takeUnless { it == JSONObject.NULL })
                }
            }.getOrNull()
        }
    }
}

File: plugins/ekey_flutter_sdk/ios/Classes/EkeyFlutterSdkPlugin.swift
import Flutter
import UIKit
import EkeySDK

/// Flutter bridge to the vendored `EkeySDK.xcframework`.
///
/// Method channel `ekey_flutter_sdk`:
///   - `initiateLogin`    – starts the eKey 2.0 app-to-app login flow.
///   - `getPendingResult` – returns + clears a result that arrived while no Dart
///     call was waiting (host was suspended/killed during the app-to-app hop).
///
/// Event channel `ekey_flutter_sdk/events` streams the same result maps and
/// replays a persisted-but-unconsumed result to the first listener.
///
/// `application(_:open:options:)` forwards the `necekey://callback` redirect back
/// into EkeySDK automatically (works with `FlutterAppDelegate` and
/// `FlutterSceneDelegate`), so the host only registers the `necekey` URL scheme
/// in its `Info.plist`.
public class EkeyFlutterSdkPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private static let defaultsKey = "ekey_flutter_sdk.pending_result"

    private var pendingResult: FlutterResult?
    private var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "ekey_flutter_sdk",
            binaryMessenger: registrar.messenger()
        )
        let events = FlutterEventChannel(
            name: "ekey_flutter_sdk/events",
            binaryMessenger: registrar.messenger()
        )
        let instance = EkeyFlutterSdkPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        events.setStreamHandler(instance)
        registrar.addApplicationDelegate(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initiateLogin":
            initiateLogin(result: result)
        case "getPendingResult":
            result(Self.takePersistedResult())
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initiateLogin(result: @escaping FlutterResult) {
        pendingResult = result
        // Presenting UIKit / creating the WKWebView must happen on the main thread.
        DispatchQueue.main.async {
            guard let presenter = Self.topViewController() else {
                self.deliver(["status": "failed", "error": "noRootViewController"])
                return
            }

            Ekey.shared.initiateLogin(from: presenter) { loginResult in
                switch loginResult {
                case .completed(let redirectURL, let codeVerifier, let identity):
                    var map: [String: Any] = [
                        "status": "completed",
                        "redirectUri": redirectURL.absoluteString,
                        "codeVerifier": codeVerifier,
                        // EKYC payload — EkeySDK does the token exchange internally.
                        "claims": Self.sanitize(identity.claims),
                    ]
                    if let kyc = identity.kycData { map["kycData"] = Self.sanitize(kyc) }
                    self.deliver(map)
                case .cancelled:
                    self.deliver(["status": "cancelled"])
                case .failed(let error):
                    self.deliver(["status": "failed", "error": String(describing: error)])
                }
            }
        }
    }

    /// Persists first (survives suspension), then hands the result to whatever
    /// Dart side is alive now.
    private func deliver(_ map: [String: Any]) {
        Self.persistResult(map)
        DispatchQueue.main.async {
            let hadWaiter = self.pendingResult != nil
            let hadSink = self.eventSink != nil
            guard hadWaiter || hadSink else { return } // stays persisted for recovery

            _ = Self.takePersistedResult() // consume
            if let waiter = self.pendingResult {
                self.pendingResult = nil
                waiter(map)
            }
            self.eventSink?(map)
        }
    }

    // MARK: - FlutterStreamHandler

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        if let pending = Self.takePersistedResult() {
            events(pending)
        }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    // MARK: - App-to-app callback

    public func application(
        _ application: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any] = [:]
    ) -> Bool {
        Ekey.shared.handleOpenURL(url)
    }

    public func application(_ application: UIApplication, handleOpen url: URL) -> Bool {
        Ekey.shared.handleOpenURL(url)
    }

    // MARK: - Persistence

    /// Coerce the claims/kyc tree to plist- and codec-safe values (drop NSNull).
    private static func sanitize(_ value: Any) -> Any {
        switch value {
        case let dict as [String: Any]:
            var out: [String: Any] = [:]
            for (k, v) in dict where !(v is NSNull) { out[k] = sanitize(v) }
            return out
        case let array as [Any]:
            return array.filter { !($0 is NSNull) }.map { sanitize($0) }
        default:
            return value
        }
    }

    private static func persistResult(_ map: [String: Any]) {
        UserDefaults.standard.set(map, forKey: defaultsKey)
    }

    private static func takePersistedResult() -> [String: Any]? {
        let defaults = UserDefaults.standard
        guard let map = defaults.dictionary(forKey: defaultsKey) else { return nil }
        defaults.removeObject(forKey: defaultsKey)
        return map
    }

    // MARK: - Helpers

    private static func topViewController() -> UIViewController? {
        guard var top = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .flatMap({ $0.windows })
            .first(where: { $0.isKeyWindow })?.rootViewController
        else {
            return nil
        }
        while let presented = top.presentedViewController {
            top = presented
        }
        return top
    }
}
